README.md
fbexport
========

FBExport is a tool for importing/exporting data and executing SQL statements on Firebird , FBCopy is a tool to copy and compare data accross Firebird databases. 


FBExport       
export and import data from Firebird databases
command line and GUI version
runs on Windows and Linux
export to comma separated values (CSV) format
export as INSERT statements
use exported data in DML statements
handles NULLs and BLOBs properly
ability to execute sql scripts from a file
FBCopy       
copy and compare data between Firebird databases
runs on Windows and Linux
automatically loads tables and compares their fields
uses Foreign Keys and Checks to determine the correct order of tables
can create ALTER TABLE and CREATE TABLE scripts needed to update the destination database
HTML overview of differences in data and metadata


CHANGES
13. Nov 2009.   Added support for INSERT or UPDATE (when insert fails)
26. Jun 2009.   Fixed bug when copying blobs with Firebird 2.1
06. Mar 2009.   Fixed bug when exporting negative values like -0.01
05. Mar 2009.   FBExport and FBCopy share the same source tree and build now

16. Feb 2008.   Version 1.70 released (at http://www.firebirdfaq.org/fbexport.php)
16. Feb 2008.   Slash character is no longer removed when loading statements from file
16. Feb 2008.   Username and password are read from ISC_USER and ISC_PASSWORD enviroment variables if available
06. Aug 2007.   Added header with column names to CSV export
06. Aug 2007.   Fixed bug with copying data on Windows (reported by Daniele Henggeler)

01. Jul 2007.   Version 1.67 released
01. Jul 2007.   Added option -B to set a different separator for CSV export (instead of comma)
01. Jul 2007.   Fixed hang with INSERT statement export when field starts with single-quote
01. Jul 2007.   Fixed hang with CSV export when field starts with quote
15. Jun 2006.   Asterisk character is no longer removed when loading statements from file
20. Mar 2006.   NUMERIC and DECIMAL columns don't suffer from conversions to doubles anymore

11. Mar 2006.   Version 1.65 released (command-line only)
09. Mar 2006.   FBExport source is now available in repository: http://svn.sourceforge.net/svnroot/fbexport
09. Mar 2006.   If charset isn't specified on command-line, FBExport uses database's charset
09. Mar 2006.   Added option -O to specify a role when connecting
09. Mar 2006.   Introduced IBPP 2.5 into sources
08. Mar 2006.   Added option -M to commit at each checkpoint

04. Mar 2005.   Version 1.60 released
04. Mar 2005.   Fixed multiple bugs in parametrised import feature (-if option).
04. Mar 2005.   In case of errors, nulls are printed as [null] instead of empty string.
06. Feb 2005.   Executing of multiple statements from file (coded by Alex Edelev).

20. Jan 2005.   Version 1.55 released
24. Nov 2004.   "If" option, user can specify full insert string with parameter list.
                It can also be used for update/delete,execute SP, not just inserts.
18. Nov 2004.   Istvan Matyasi joined the project
30. Sep 2004.   Fixed bug: program would crash if checkpoint count was set to zero
17. Sep 2004.   Fixed minor bug in Linux GUI version: charset wasn't loaded right

05. Sep 2004.   Version 1.51 released
05. Sep 2004.   Linux GUI binary is part of main package from now on
09. Aug 2004.   Binaries compiled against IBPP v2.3 which has much better support for Firebird 1.5
06. Aug 2004.   CSV exported values now have double-quotes escaped, i.e. " in value becomes ""

10. May 2004.   Version 1.50 released
08. May 2004.   All changes since 1.45 are now also in GUI version
02. May 2004.   Wrote nice HTML manual that explains usage and options
02. May 2004.   Changed -E option default value to zero. To allow unlimited number of errors use value -1.
02. May 2004.   Data can be exported as INSERT statemens
02. May 2004.   Fixed minor memory leak (few bytes) when importing blobs with errors.
02. May 2004.   Fixed bug: input file wasn't closed at end
02. May 2004.   Input can be read from stdin, thus data-pumping can be done with simple piping of 2 fbexport commands
30. Apr 2004.   Output that is not data, but information about what program is doing, is sent to stderr
30. Apr 2004.   Output (data) can be redirected to stdout instead of file.
30. Apr 2004.   User can select custom date and time formatting for CSV export
30. Apr 2004.   Data can be exported in CSV (comma separated values) format
29. Apr 2004.   Support for no_auto_undo transaction option which makes Firebird use less memory for operation

26. Mar 2004.   Version 1.45 released
25. Mar 2004.   "Verbatim copy" option. FBExport builds list of columns in alphabetical order (-V option)
25. Mar 2004.   Added support to specify charset when connecting (-A option)

27. Feb 2004.   Version 1.43 released
26. Feb 2004.   Fixed bug: importing of smallint didn't work. (Thanx to Ed Epstein for finding it)

10. Feb 2004.   Version 1.42 released
09. Feb 2004.   GUI version reloads the previous settings at start
09. Feb 2004.   Added support for Int64 (decimal(18,0))

14. Jan 2004.   Version 1.40 released
13. Jan 2004.   Finished first complete version of FBExport GUI
12. Jan 2004.   Created *real* Makefile for Linux (instead the earlier one-liner)
03. Jan 2004.   Separated code for startup and printout so the GUI can use the same code as the command-line version
31. Dec 2003.   Changed some of the code to support Dialect 1
30. Dec 2003.   No need to separate options from values, you can write: -Usysdba or -U sysdba
30. Dec 2003.   FBExport shows which version of IBPP it is using

29. Dec 2003.   Version 1.35 released
27. Dec 2003.   Added support for BLOB datatype
26. Dec 2003.   Username defaults to SYSDBA to make DBA's life easier ;)
19. Dec 2003.   Added "make static" option to Linux Makefile to build executable with IBPP linked in it.
21. Nov 2003.   Added option (-L) to show list of connected users

03. Nov 2003.   Version 1.30 released
03. Nov 2003.   Linked with IBPP 2.2.2.3, so wider date range is supported: from 01 Jan 0001 to 31 Dec 9999.
01. Nov 2003.   Added support for (var)chars longer than 255 characters
01. Nov 2003.   Added support for time datatype
31. Oct 2003.   Host defaults to LOCALHOST so it's not a required value anymore
31. Oct 2003.   Command line flags are not case-sensitive anymore
31. Oct 2003.   Fixed differences between NULL values and empty varchars

08. Avg 2003.   Version 1.20 released
08. Avg 2003.   The package is from now on released as a single file: fbx000-full.zip
                where 000 is the version number.
07. Avg 2003.   Fixed some bugs, and changed the code to use version 2.x of IBPP library

25. Dec 2002.   Version 1.00 released
24. Dec 2002.   Removed all hash-mark idea from arguments class

ibpp/date.cpp
///////////////////////////////////////////////////////////////////////////////
//
//	File    : $Id: date.cpp 27 2006-05-20 19:31:15Z mbabuskov $
//	Subject : IBPP, Date class implementation
//
///////////////////////////////////////////////////////////////////////////////
//
//	(C) Copyright 2000-2006 T.I.P. Group S.A. and the IBPP Team (www.ibpp.org)
//
//	The contents of this file are subject to the IBPP License (the "License");
//	you may not use this file except in compliance with the License.  You may
//	obtain a copy of the License at http://www.ibpp.org or in the 'license.txt'
//	file which must have been distributed along with this file.
//
//	This software, distributed under the License, is distributed on an "AS IS"
//	basis, WITHOUT WARRANTY OF ANY KIND, either express or implied.  See the
//	License for the specific language governing rights and limitations
//	under the License.
//
///////////////////////////////////////////////////////////////////////////////
//
//	COMMENTS
//	* Tabulations should be set every four characters when editing this file.
//
///////////////////////////////////////////////////////////////////////////////

#ifdef _MSC_VER
#pragma warning(disable: 4786 4996)
#ifndef _DEBUG
#pragma warning(disable: 4702)
#endif
#endif

#include "_ibpp.h"

#ifdef HAS_HDRSTOP
#pragma hdrstop
#endif

#include <time.h>		// Can't use <ctime> thanks to MSVC6 buggy library

using namespace ibpp_internals;

void IBPP::Date::Today()
{
	time_t systime = time(0);
	tm* loctime = localtime(&systime);

	if (! IBPP::itod(&mDate, loctime->tm_year + 1900,
		loctime->tm_mon + 1, loctime->tm_mday))
			throw LogicExceptionImpl("Date::Today", _("Out of range"));
}

void IBPP::Date::SetDate(int dt)
{
	if (! IBPP::dtoi(dt, 0, 0, 0))
		throw LogicExceptionImpl("Date::SetDate", _("Out of range"));
	mDate = dt;
}

void IBPP::Date::SetDate(int year, int month, int day)
{
	if (! IBPP::itod(&mDate, year, month, day))
		throw LogicExceptionImpl("Date::SetDate", _("Out of range"));
}

void IBPP::Date::GetDate(int& year, int& month, int& day) const
{
	if (! IBPP::dtoi(mDate, &year, &month, &day))
		throw LogicExceptionImpl("Date::GetDate", _("Out of range"));
}

int IBPP::Date::Year() const
{
	int year;
	if (! IBPP::dtoi(mDate, &year, 0, 0))
		throw LogicExceptionImpl("Date::Year", _("Out of range"));
	return year;
}

int IBPP::Date::Month() const
{
	int month;
	if (! IBPP::dtoi(mDate, 0, &month, 0))
		throw LogicExceptionImpl("Date::Month", _("Out of range"));
	return month;
}

int IBPP::Date::Day() const
{
	int day;
	if (! IBPP::dtoi(mDate, 0, 0, &day))
		throw LogicExceptionImpl("Date::Day", _("Out of range"));
	return day;
}

void IBPP::Date::Add(int days)
{
	int newdate = mDate + days;		// days can be signed
	if (! IBPP::dtoi(newdate, 0, 0, 0))
		throw LogicExceptionImpl("Date::Add()", _("Out of range"));
	mDate = newdate;
}

void IBPP::Date::StartOfMonth()
{
	int year, month;
	if (! IBPP::dtoi(mDate, &year, &month, 0))
		throw LogicExceptionImpl("Date::StartOfMonth()", _("Out of range"));
	if (! IBPP::itod(&mDate, year, month, 1))		// First of same month
		throw LogicExceptionImpl("Date::StartOfMonth()", _("Out of range"));
}

void IBPP::Date::EndOfMonth()
{
	int year, month;
	if (! IBPP::dtoi(mDate, &year, &month, 0))
		throw LogicExceptionImpl("Date::EndOfMonth()", _("Out of range"));
	if (++month > 12) { month = 1; year++; }
	if (! IBPP::itod(&mDate, year, month, 1))	// First of next month
		throw LogicExceptionImpl("Date::EndOfMonth()", _("Out of range"));
	mDate--;	// Last day of original month, all weird cases accounted for
}

IBPP::Date::Date(int year, int month, int day)
{
	SetDate(year, month, day);
}

IBPP::Date::Date(const IBPP::Date& copied)
{
	mDate = copied.mDate;
}

IBPP::Date& IBPP::Date::operator=(const IBPP::Timestamp& assigned)
{
	mDate = assigned.GetDate();
	return *this;
}

IBPP::Date& IBPP::Date::operator=(const IBPP::Date& assigned)
{
	mDate = assigned.mDate;
	return *this;
}

// The following date calculations were inspired by web pages found on
// Peter Baum web homepage at 'http://www.capecod.net/~pbaum/'.
// His contact info is at : 'http://home.capecod.net/~pbaum/contact.htm'.
// Please, understand that Peter Baum is not related to this IBPP project.
// So __please__, do not contact him regarding IBPP matters.

//	Take a date, in its integer format as used in IBPP internals and splits
//	it in year (4 digits), month (1-12), day (1-31)

bool IBPP::dtoi (int date, int *y, int *m, int *d)
{
    int RataDie, Z, H, A, B, C;
    int year, month, day;

	// Validity control.
	if (date < IBPP::MinDate || date > IBPP::MaxDate)
		return false;

	// The "Rata Die" is the date specified as the number of days elapsed since
	// 31 Dec of year 0. So 1 Jan 0001 is 1.

	RataDie = date + ibpp_internals::consts::Dec31_1899;	// Because IBPP sets the '0' on 31 Dec 1899.

    Z = RataDie + 306;
    H = 100*Z - 25;
    A = H/3652425;
    B = A - A/4;
    year = (100*B + H) / 36525;
    C = B + Z - 365*year - year / 4;
    month = (5*C + 456) / 153;
    day = C - (153*month - 457) / 5;
    if (month > 12) { year += 1; month -= 12; }

	if (y != 0) *y = (int)year;
	if (m != 0) *m = (int)month;
	if (d != 0) *d = (int)day;

	return true;
}

//	Take a date from its components year, month, day and convert it to the
//	integer representation used internally in IBPP.

bool IBPP::itod (int *pdate, int year, int month, int day)
{
    int RataDie, result;
	int y, m, d;

	d = day;	m = month;		y = year;
    if (m < 3) { m += 12; y -= 1; }
    RataDie = d + (153*m - 457) / 5 + 365*y + y/4 - y/100 + y/400 - 306;

	result = RataDie - ibpp_internals::consts::Dec31_1899;   // Because IBPP sets the '0' on 31 Dec 1899

	// Validity control
	if (result < IBPP::MinDate || result > IBPP::MaxDate)
		return false;

	*pdate = result;
	return true;
}

//	Eof
